examples/by_feature/gradient_accumulation_for_autoregressive_models.py

```diff
@@ -0,0 +1,341 @@
+
+            # Sync gradients and perform optimization steps once every gradient_accumulation_steps
+            grad_norm = accelerator.clip_grad_norm_(model.parameters(), max_grad_norm)
+            optimizer.step()
+            lr_scheduler.step()
+            optimizer.zero_grad()
+
+            losses = accelerator.gather(sum(losses)).sum().item() / (
+                accelerator.num_processes * gradient_accumulation_steps
+            )
+
+            grad_norm = grad_norm.detach().item() if isinstance(grad_norm, torch.Tensor) else grad_norm
+            accelerator.print(
+                f"epoch {epoch} - update step {update_step}:: grad norm: {grad_norm} ::train loss: {losses}"
+            )
+            if args.with_wandb_tracking:
+                accelerator.log(
+                    {
+                        "train/grad_norm": grad_norm,
+                        "train/epoch": epoch,
+                        "train/loss": losses,
+                    },
+                    step=update_step + total_gradient_updates * epoch,
+                )
+        model.eval()
+        losses = []
+        for step, batch in enumerate(eval_dataloader):
+            with torch.no_grad():
+                outputs = model(**batch, use_cache=False)
+            eval_loss = outputs.loss
+            losses.append(accelerator.gather_for_metrics(loss.repeat(EVAL_BATCH_SIZE)))
+
+        losses = torch.cat(losses)
+        try:
+            eval_loss = torch.mean(losses)
+            perplexity = math.exp(eval_loss)
+        except OverflowError:
+            perplexity = float("inf")
+
+        # Use accelerator.print to print only on the main process.
+        accelerator.print(f"epoch {epoch}:: eval perplexity: {perplexity} eval_loss: {eval_loss}")
+        if args.with_wandb_tracking:
+            accelerator.log(
+                {
+                    "eval/perplexity": perplexity,
+                    "eval/loss": eval_loss,
+                    "eval/epoch": epoch,
+                },
+                step=update_step + total_gradient_updates * epoch,
+            )
+    accelerator.end_training()
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Simple example of training script.")
+    parser.add_argument(
+        "--mixed_precision",
+        type=str,
+        default=None,
+        choices=["no", "fp16", "bf16", "fp8"],
+        help="Whether to use mixed precision. Choose"
+        "between fp16 and bf16 (bfloat16). Bf16 requires PyTorch >= 1.10."
+        "and an Nvidia Ampere GPU.",
+    )
+
+    parser.add_argument(
+        "--gradient_accumulation_steps",
+        type=int,
+        default=1,
+        help="The number of minibatches to be ran before gradients are accumulated.",
+    )
+    parser.add_argument(
+        "--per_device_batch_size",
+        type=int,
+        default=2,
+        help="The number of minibatches to be ran before gradients are accumulated.",
```

Comment:
Shouldn't this be "The size of each minibatch"?
```suggestion
        help="The size of each minibatch",
```